rootston/output.c

```diff
@@ -407,13 +407,73 @@ static void set_mode(struct wlr_output *output,
 	}
 }
 
+static void update_output_manager_config(struct roots_desktop *desktop) {
+	struct wlr_output_configuration_v1 *config =
+		wlr_output_configuration_v1_create();
+
+	struct roots_output *output;
+	wl_list_for_each(output, &desktop->outputs, link) {
+		wlr_output_configuration_head_v1_create(config, output->wlr_output);
+	}
+
+	wlr_output_manager_v1_set_configuration(desktop->output_manager_v1, config);
+}
+
+void handle_output_manager_apply(struct wl_listener *listener, void *data) {
+	struct roots_desktop *desktop =
+		wl_container_of(listener, desktop, output_manager_apply);
+	struct wlr_output_configuration_v1 *config = data;
+
+	bool ok = true;
+	struct wlr_output_configuration_head_v1 *config_head;
+	wl_list_for_each(config_head, &config->heads, link) {
+		struct wlr_output *wlr_output = config_head->state.output;
+		ok &= wlr_output_enable(wlr_output, config_head->state.enabled);
+		if (!config_head->state.enabled) {
+			continue;
+		}
+		if (config_head->state.mode != NULL) {
+			ok &= wlr_output_set_mode(wlr_output, config_head->state.mode);
+		} else {
+			ok &= wlr_output_set_custom_mode(wlr_output,
+				config_head->state.custom_mode.width,
+				config_head->state.custom_mode.height,
+				config_head->state.custom_mode.refresh);
+		}
+		wlr_output_layout_add(desktop->layout, wlr_output,
+			config_head->state.x, config_head->state.y);
+		wlr_output_set_transform(wlr_output, config_head->state.transform);
+		wlr_output_set_scale(wlr_output, config_head->state.scale);
+	}
+
+	if (ok) {
+		wlr_output_configuration_v1_send_succeeded(config);
+	} else {
+		wlr_output_configuration_v1_send_failed(config);
+	}
+	wlr_output_configuration_v1_destroy(config);
+
+	update_output_manager_config(desktop);
```

Comment:
This is going to leave outputs in a "random" state (ie. a state that works but is only a subset of the requested state).

A better way to apply the configuration would be to first disable what needs to be disabled, then enable what needs to be enabled. This would ensure we don't fail if we first try to enable more outputs than we can and then disable some (even if in the end with the current wlroots API everything will be fine as we enable outputs that want to when disabling other outputs).

I'd like to avoid anything more complex than this (e.g. trying to restore the state) because it'll go away when we have proper atomic support in `wlr_backend`.

Reply:
Do we have a bug for "proper atomic support in `wlr_backend`". If not could you file one? The closest thing I've seen is https://github.com/swaywm/wlr-protocols/pull/38#issuecomment-470472049

Reply:
https://github.com/swaywm/wlroots/issues/1640